src/Http/Routing/src/RouteOptions.cs

```diff
@@ -89,6 +90,15 @@ internal ICollection<EndpointDataSource> EndpointDataSources
     /// </summary>
     internal IDictionary<string, Type> TrimmerSafeConstraintMap => _constraintTypeMap;
 
+    /// <summary>
+    /// Add Regex-based constraints to the constraint map (e.g. alpha and regex constraint types). This is called automatically by <see cref="RoutingServiceCollectionExtensions.AddRouting(IServiceCollection)"/>.
+    /// </summary>
+    public void AddRegexConstraint()
+    {
+        // Regex-based constraints
+        AddConstraint<RegexInlineRouteConstraint>(_constraintTypeMap, "regex");
+    }
```

Comment:
Just throwing this out for everyone's consideration:

Would it be more understandable if we just dropped all the built-in route constraints when using `CreateSlimBuilder()`? Then apps could either opt into adding them all back with a `AddDefaultConstraints()` API, or one-by-one for each constraint they actually used in the app?

I kind of think that plan is a bit more understandable from someone who just walks up and sees that constraints don't work with CreateSlimBuilder out of the box. And they can opt-in to whichever they want. Where with the current proposal, it would be rather confusing - "why do all the other constraints work, but regex doesn't?"

Reply:
I think we should register them. It makes it more likely apps just work with slim builder, and we can provide a very specific error message with instructions for adding regex via https://github.com/dotnet/aspnetcore/pull/46227#issuecomment-1404489429

Reply:
@JamesNK, latest update includes your suggestion above. One discussion point, I've aimed to make this new route constraint internal which means I need to make the policy activator search for internal constructors on types. This doesn't seem particularly safe but also don't want to leak this constraint into the public API.